src/operators.jl

```diff
@@ -363,201 +363,163 @@ function Compat.LinearAlgebra.diagm(x::AbstractVector{<:AbstractVariableRef})
-    for i ∈ 1:m, j ∈ 1:s
-        q = ret[i,j]
-        _sizehint_expr!(q, n)
-        for k ∈ 1:n
-            tmp = convert(T, lhs[i,k]*rhs[k,j])
-            add_to_expression!(q, tmp)
-        end
+###############################################################################
+# Matrix/Vector Arithmetic with JuMP eltypes
+###############################################################################
+
+###############################################################################
+# convenience/utility definitions
+
+const GenericAffOrQuadExpr = Union{GenericAffExpr,GenericQuadExpr}
+
+densify_with_jump_eltype(x::AbstractMatrix) = convert(Matrix, x)
+
+function densify_with_jump_eltype(x::SparseMatrixCSC{V}) where V<:AbstractVariableRef
+    return convert(Matrix{GenericAffExpr{Float64,V}}, x)
+end
+
+# see https://github.com/JuliaLang/julia/pull/18218
+_A_mul_B_eltype(::Type{R}, ::Type{S}) where {R,S} = typeof(one(R) * one(S) + one(R) * one(S))
+
+_A_mul_B_ret_dims(A::AbstractMatrix, B::AbstractVector) = (size(A, 1),)
+_A_mul_B_ret_dims(A::AbstractMatrix, B::AbstractMatrix) = (size(A, 1), size(B, 2))
+
+# don't do size checks here; defer that to `_A_mul_B(A, B)`
+function _A_mul_B_ret(A, B, dims...)
+    T = _A_mul_B_eltype(eltype(A), eltype(B))
+    ret = Array{T}(undef, _A_mul_B_ret_dims(A, B))
+    return fillwithzeros!(ret, T)
+end
+
+if VERSION < v"0.7-"
+    _At_mul_B_ret_dims(A::AbstractMatrix, B::AbstractVector) = (size(A, 2),)
+    _At_mul_B_ret_dims(A::AbstractMatrix, B::AbstractMatrix) = (size(A, 2), size(B, 2))
+    function _At_mul_B_ret(A, B, dims...)
+        T = _A_mul_B_eltype(eltype(A), eltype(B))
+        ret = Array{T}(undef, _At_mul_B_ret_dims(A, B))
+        return fillwithzeros!(ret, T)
     end
-    ret
 end
 
-# this computes lhs.'*rhs and places it in ret
-function _multiplyt!(ret::Array{T}, lhs::Array, rhs::Array) where T<:JuMPTypes
-    m, n = size(lhs,2), size(lhs,1) # transpose
-    r, s = size(rhs,1), size(rhs,2)
-    for i ∈ 1:m, j ∈ 1:s
-        q = ret[i,j]
-        _sizehint_expr!(q, n)
-        for k ∈ 1:n
-            tmp = convert(T, lhs[k,i]*rhs[k,j]) # transpose
+function fillwithzeros!(A, ::Type{T}) where {T}
+    for I in eachindex(A)
+        A[I] = zero(T)
+    end
+    return A
+end
+
+###############################################################################
+# `_A_mul_B!(ret, A, B)` adds the result of `A*B` into the buffer `ret`. We roll our own
+# matmul here (instead of using Julia's generic fallbacks) because doing so allows us to
+# accumulate the expressions for the inner loops in-place. Additionally, Julia's generic
+# fallbacks can be finnicky when your array elements aren't `<:Number`.
+#
+# No bounds/size checks are performed; it is expected that the caller has done this, has
+# ensured that the eltype of `ret` is appropriate, and has zeroed the elements of `ret` (if
+# desired).
+
+function _A_mul_B!(ret::AbstractArray{T}, A, B) where {T<:JuMPTypes}
```

Comment:
Nit: Spaces between binary operators, so `T <: JuMPTypes`.

Reply:
Hmmm. No spaces surrounding `<:` seems to be the norm in the rest of the codebase. Seems best to change them all at once or not all to avoid inconsistent style - should I do that in this PR or leave until later?

Reply:
We're fixing the style issues as we touch lines of code. It's pretty painful to update the whole code base at once, but you're welcome to if you'd like.

Reply:
This seems like an easy one, I'll just go for it